File: 1114-binary-search-tree-to-greater-sum-tree/binary-search-tree-to-greater-sum-tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    private int globalSum;
    public TreeNode BstToGst(TreeNode root) {
        this.globalSum = BstGetSum(root);
        return BstToGstAcc(root);
    }

    private TreeNode BstToGstAcc(TreeNode root)
    {
        if (root == null)
            return root;

        BstToGstAcc(root.left);

        // inorder traversal logic here
        this.globalSum -= root.val;
        root.val += this.globalSum;

        BstToGstAcc(root.right);
        
        return root;
    }

    private int BstGetSum(TreeNode root)
    {
        return (root == null) ? 0 : root.val + BstGetSum(root.left) + BstGetSum(root.right);
    }
}

File: 1114-binary-search-tree-to-greater-sum-tree/binary-search-tree-to-greater-sum-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int globalSum;
int bstGetSum(struct TreeNode* root)
{
    return (root == NULL) ? 0 : root->val + bstGetSum(root->left) + bstGetSum(root->right);
}

struct TreeNode* bstToGstAcc(struct TreeNode* root)
{
    if (root == NULL)
        return root;
    
    // inorder
    root->left = bstToGstAcc(root->left);

    globalSum -= root->val;
    root->val += globalSum;

    root->right = bstToGstAcc(root->right);

    return root;
}

struct TreeNode* bstToGst(struct TreeNode* root) {
    globalSum = bstGetSum(root);
    return bstToGstAcc(root);
}



